### PR TITLE
Add #[make_varule] proc macro

### PR DESCRIPTION
**Pull request by Manishearth**
Built on https://github.com/unicode-org/icu4x/pull/1603

Part of #1079

This does not implement ZCF yet or handle custom last fields, I plan to handle those later.

### CODE REVIEW

File: utils/zerovec/derive/src/varule.rs [hunk 1]
```diff
@@ -110,3 +114,313 @@ pub fn derive_impl(input: &DeriveInput) -> TokenStream2 {
...
+                let unaligned = zerovec::ule::AsULE::as_unaligned(self.#accessor);
+                let unaligned_slice = &[unaligned];
+                let src = <#ty as zerovec::ule::ULE>::as_byte_slice(unaligned_slice);
+                out.copy_from_slice(src);
+            )
+        });
+
+    let last_bytes = last_field_info.encode_func(quote!(self.#last_field_name));
+    quote!(
+        unsafe impl #maybe_lt_bound zerovec::ule::custom::EncodeAsVarULE<#ule_name> for #name #maybe_lt_bound {
```
Comment (sffc) [inline utils/zerovec/derive/src/varule.rs:330]: Nit: Safety comment

**Comment by Manishearth** [comment]
Ready for review!

File: utils/zerovec/derive/src/varule.rs [hunk 2]
```diff
@@ -110,3 +114,399 @@ pub fn derive_impl(input: &DeriveInput) -> TokenStream2 {
...
+                        for param in &params.args {
+                            match param {
+                                GenericArgument::Lifetime(ref lt) if lifetime.is_none() => {
+                                    lifetime = Some(lt)
+                                }
+                                GenericArgument::Type(ref ty) if generic.is_none() => {
+                                    generic = Some(ty)
+                                }
+                                _ => return Err("Can only automatically detect corresponding VarULE types for path \
+                                                 types with at most one lifetime and at most one generic parameter".into()),
```
Comment (sffc) [inline utils/zerovec/derive/src/varule.rs:405]:
```suggestion
                                                 types with at most one lifetime and exactly one generic parameter".into()),
```

File: utils/zerovec/derive/src/varule.rs [hunk 3]
```diff
@@ -110,3 +114,399 @@ pub fn derive_impl(input: &DeriveInput) -> TokenStream2 {
...
+                            }
+                        }
+
+                        // Must be exactly one generic parameter
+                        let generic = if let Some(g) = generic {
+                            g
+                        } else {
+                            return Err("Can only automatically detect corresponding VarULE types for path \
+                                        types with at most one lifetime and at most one generic parameter".into());
```
Comment (sffc) [inline utils/zerovec/derive/src/varule.rs:414]:
```suggestion
                                        types with at most one lifetime and exactly one generic parameter".into());
```

File: utils/zerovec/derive/src/varule.rs [hunk 4]
```diff
@@ -110,3 +114,399 @@ pub fn derive_impl(input: &DeriveInput) -> TokenStream2 {
...
+                                path types with a single path segment".into());
+                }
+                let segment = typath.path.segments.first().unwrap();
+                match segment.arguments {
+                    PathArguments::None => {
+                        if segment.ident == "String" {
+                            Ok(LastField::Growable(OwnULETy::Str))
+                        } else {
+                            Err("Can only automatically detect corresponding VarULE types for path types \
+                                        that are not Cow, ZeroVec, VarZeroVec, Box, String, or Vec".into())
```
Comment (sffc) [inline utils/zerovec/derive/src/varule.rs:389]:
```suggestion
                                        that are Cow, ZeroVec, VarZeroVec, Box, String, or Vec".into())
```

File: utils/zerovec/derive/src/varule.rs [hunk 5]
```diff
@@ -110,3 +114,399 @@ pub fn derive_impl(input: &DeriveInput) -> TokenStream2 {
...
+
+                        let ident = segment.ident.to_string();
+
+                        if lifetime.is_some() {
+                            match &*ident {
+                                "ZeroVec" => Ok(LastField::ZeroVec(generic)),
+                                "VarZeroVec" => Ok(LastField::VarZeroVec(generic)),
+                                "Cow" => OwnULETy::new(generic, "Cow").map(LastField::Cow),
+                                _ => Err("Can only automatically detect corresponding VarULE types for path \
+                                          types that are not Cow, ZeroVec, VarZeroVec, Box, String, or Vec".into()),
```
Comment (sffc) [inline utils/zerovec/derive/src/varule.rs:425]:
```suggestion
                                          types that are Cow, ZeroVec, VarZeroVec, Box, String, or Vec".into()),
```

File: utils/zerovec/derive/src/varule.rs [hunk 6]
```diff
@@ -110,3 +114,399 @@ pub fn derive_impl(input: &DeriveInput) -> TokenStream2 {
...
+                            }
+                        } else {
+                            match &*ident {
+                                "Vec" => Ok(LastField::Growable(OwnULETy::Slice(generic))),
+                                "Box" => OwnULETy::new(generic, "Box").map(LastField::Boxed),
+                                _ => Err("Can only automatically detect corresponding VarULE types for path types \
+                                          that are not Cow, ZeroVec, VarZeroVec, Box, String, or Vec".into()),
```
Comment (sffc) [inline utils/zerovec/derive/src/varule.rs:432]:
```suggestion
                                          that are Cow, ZeroVec, VarZeroVec, Box, String, or Vec".into()),
```

File: utils/zerovec/derive/src/varule.rs (hunk 2 above)
Comment (Manishearth) [reply utils/zerovec/derive/src/varule.rs]: No, `str` is a path type without one generic

File: utils/zerovec/derive/src/varule.rs (hunk 3 above)
Comment (Manishearth) [reply utils/zerovec/derive/src/varule.rs:414]: No, `str` is a path type without one generic

File: utils/zerovec/src/varzerovec/slice.rs [hunk 7]
```diff
@@ -360,6 +360,28 @@ where
...
 
+impl<T> Eq for VarZeroSlice<T>
+where
+    T: VarULE,
+    T: ?Sized,
+    T: Eq,
+{
+}
+
+impl<T: VarULE + ?Sized + PartialOrd> PartialOrd for VarZeroSlice<T> {
```
Comment (sffc) [inline utils/zerovec/src/varzerovec/slice.rs:371]: Question: Where do you use these impls?  If they aren't needed for this PR, consider moving them to their own PR

**Review by sffc — APPROVED** [review]

File: utils/zerovec/src/varzerovec/slice.rs (hunk 7 above)
Comment (Manishearth) [reply utils/zerovec/src/varzerovec/slice.rs:371]: They're needed so that we can `#[derive(Ord)]` etc as necessary

**Review by sffc — APPROVED** [review]
